src/SharpLearning.Backend/TrainingApi.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SharpLearning.Backend
+{
+    class TrainingApi
+    {
+        interface IDataSource
+        {
+            Tuple<ITensorSymbol, ITensorSymbol> NextBatch(int size); // Return type should probably be different.
+        }
+
+        // training API brainstorm
+        interface IOptimizer
+        {
+            ITensorSymbol Minimize(ITensorSymbol cost);
+        };
+
+        interface ILoss
+        {
+            ITensorSymbol Loss(ITensorSymbol targets, ITensorSymbol predictions);
+        }
+
+        void Suggestion1()
+        {
+            // This will probably not be possible with current CNTK and TensorFlowSharp C# API state.
+            // Might also increase back and forth copying between CPU/GPU.
+
+            // var graph = DefineGraph(int[] shape, IBackend backend)
+            // var loss = new CrossEntropy();
+            // var metric = new ClassificationAccuracy();
+            // var optimizer = new SGD(learningRate: 0.001);
+
+            //using (var session = StartSession(graph))            
+            // for epochs
+
+            //    reset/shuffle datasource
+
+            //    datasource prefetcher
+
+            //    for iterations
+            //          Tuple<ITensorSymbol> batch = source.NextBatch(batchSize: 64); // load next minibatch
```

Comment:
We have to be very careful here... symbols cannot be direct representations of loaded data. This is important. Symbols are not actual "instances". TF couples instances with symbols by coupling placeholder symbols to actual instances if I remember correctly.

The same applies to all the below. The suggestion is based on symbols, but it is unclear whether the suggestion is actually intending to use concrete instances or define some kind of operation via symbols.

No doubt this will get more concrete, when we start actually working on it. The notion of forward/backward is interesting, although I have not seen this divide explicitly in TF for example. 

Also, don't like "Tuple<>" use ;)

Reply:
Yes, this is rather superficial, but does serve as a starting point to chose a direction from. When we start to implement types, it will become more concrete and then we can decide which parts should be handled via symbols and which part should be concrete instances.

I agree with the use of Tuple<>, it is just a placeholder until we decide on the real types.